### DISCUSSION TITLE
Polymorphism joins against all sub types when querying

### DISCUSSION THREAD

**Discussion by nfplee** [nhibernate/nhibernate-core#3163] [Q&A] [answered]
My application was previously structured like so:

```cs
public abstract class Category : Entity {
    public virtual Category? Parent { get; set; }
    ...
}

public abstract class ContentCategory : Category { ... }
public class BlogPostCategory : ContentCategory { ... }
public class PageCategory : ContentCategory { ... }

public abstract class ContentItem : Entity {
    public virtual ContentCategory Category { get; set; }
    ...
}

public BlogPost : ContentItem { ... }
public Page : ContentItem { ... }
```

Where each type (apart from `Entity`) is mapped to a table in the database.

However if I tried retrieving the `BlogPost` or `Page` and fetching the `Category`, like so:

```cs
await session.Query<BlogBlog>().Fetch(p => p.Category).ToListAsync();
```

I noticed that the SQL does a join against the `PageCategory` aswell as the `BlogPostCategory`. I understand this is because the `Category` property is of type `ContentCategory` and therefore it doesn't understand that it should be a `BlogPostCategory`. I decided to take the `Category` property out of the `ContentItem` type and move it to under the `BlogPost` and `Page` types, like so:

```cs
public BlogPost : ContentItem {
    public virtual BlogPostCategory Category { get; set; }
    ...
}

public Page : ContentItem {
    public virtual PageCategory Category { get; set; }
    ...
}
```

This fixed the issue. However if I now say:

```cs
await session.Query<BlogBlogCategory>().Fetch(c => c.Parent).ToListAsync();
```

The same problem exists as the `Parent` property is of type `Category` and not `BlogPostCategory`. I don't like the idea of moving the `Parent` property out of the `Category` type as I did to solve the previous issue. I also can't see how I can solve this with generics because if I make the `Category` type generic (to pass the `Parent` type) then I won't be able to map the `Category` type. Finally after reading up on this I tried adding `Polymorphism(PolymorphismType.Explicit)` to my `Category` mapping but this didn't seem to change anything.

I'd really appreciate any advice.

**Comment by bahusoid** [comment, marked as answer]
Mappings?

**Reply by nfplee** [reply]
Anything type which inherits from `Entity` is mapped as a class mapping and all other types use a joined subclass mapping. Here's a couple of examples:

```cs
public class ContentItemMapping : ClassMapping<ContentItem> {
    public ContentItemMapping() {
        Table("ContentItems");
        Id(x => x.Id, m => m.Generator(NHibernate.Mapping.ByCode.Generators.Identity));
        ...
    }
}

public class BlogPostMapping : JoinedSubclassMapping<BlogPost> {
    public BlogPostMapping() {
        Table("BlogPosts");
        Key(k => k.Column("Id"));
        ManyToOne(x => x.Category);
        ...
    }
}
```

**Reply by bahusoid** [reply]
I would avoid subclass mapping if possible.

And regarding your issues. You can keep `ContentItem.Category` in base class as `ContentCategory`. But instead of mapping it in `ContentItemMapping` - map it in concrete subclass mappings with specifying concrete type inside `ManyToOne` call.
```C#
public class ContentItemMapping : ClassMapping<ContentItem> {
    public ContentItemMapping() {
        Table("ContentItems");
        Id(x => x.Id, m => m.Generator(NHibernate.Mapping.ByCode.Generators.Identity));
        // ManyToOne(x => x.Category); <-- Remove Category mapping
        ...
    }
}

public class BlogPostMapping : JoinedSubclassMapping<BlogPost> {
    public BlogPostMapping() {
        Table("BlogPosts");
        Key(k => k.Column("Id"));
        ManyToOne(x => x.Category, m => m.Class(typeof(BlogPostCategory ))); <-- specify concrete type via Class method
        ...
    }

public abstract class ContentItem : Entity {
    public virtual ContentCategory Category { get; set; }
    ...
}
```

Same applies for `Parent` property.

**Reply by nfplee** [reply]
Thanks, I never even thought of not mapping it in the base class. I think I'll keep my `ContentItem` as is, since I think making the category optional against the content types is a good thing. However this should point me in the right direction for the `Parent` property.
